alchemy/src/aws/control/client.ts

```diff
@@ -0,0 +1,385 @@
+    patchDocument: Record<string, any>
+  ): Promise<ProgressEvent> {
+    return this.sync("UpdateResource", {
+      TypeName: typeName,
+      Identifier: identifier,
+      PatchDocument: JSON.stringify(patchDocument),
+    });
+  }
+
+  /**
+   * Delete a resource
+   */
+  public async deleteResource(
+    typeName: string,
+    identifier: string
+  ): Promise<ProgressEvent> {
+    return this.sync("DeleteResource", {
+      TypeName: typeName,
+      Identifier: identifier,
+    });
+  }
+
+  /**
+   * List resources of a given type
+   */
+  public listResources(typeName: string, nextToken?: string) {
+    return this.fetch<{
+      ResourceDescriptions: Array<{
+        Identifier: string;
+        Properties: Record<string, any>;
+      }>;
+      NextToken?: string;
+    }>("ListResources", {
+      TypeName: typeName,
+      NextToken: nextToken,
+    });
+  }
+
+  public async sync(action: string, props?: any): Promise<ProgressEvent> {
+    const { ProgressEvent } = await this.fetch<{
+      ProgressEvent: ProgressEvent;
+    }>(action, props);
+
+    let attempts = 0;
+    let delay = this.initialPollingDelay;
+    const startTime = Date.now();
+
+    while (attempts < this.maxPollingAttempts) {
+      if (Date.now() - startTime > this.operationTimeout) {
+        throw new TimeoutError(
+          `Operation ${ProgressEvent.Operation} timed out after ${
+            this.operationTimeout / 1000
+          } seconds`
+        );
+      }
+
+      try {
+        const response = await this.fetch<{
+          ProgressEvent: ProgressEvent;
+        }>("GetResourceRequestStatus", {
+          RequestToken: ProgressEvent.RequestToken,
+        });
+
+        if (response.ProgressEvent.OperationStatus === "SUCCESS") {
+          return response.ProgressEvent;
+        }
+
+        if (response.ProgressEvent.OperationStatus === "FAILED") {
+          if (response.ProgressEvent.ErrorCode === "AlreadyExists") {
+            throw new AlreadyExistsError(response.ProgressEvent);
+          }
+          throw new UpdateFailedError(response.ProgressEvent);
+        }
+
+        // Use the suggested retry delay if provided
+        const waitTime = response.ProgressEvent.RetryAfter
+          ? Math.max(0, response.ProgressEvent.RetryAfter * 1000 - Date.now())
+          : delay;
+
+        await new Promise((resolve) => setTimeout(resolve, waitTime));
+        delay = Math.min(delay * 2, this.maxPollingDelay);
+        attempts++;
+      } catch (error: any) {
+        if (error instanceof CloudControlError) {
+          throw error;
+        }
+
+        // For retryable errors, continue polling
+        // console.warn(
+        //   `Error polling operation ${operationId} (attempt ${attempts + 1}):`,
+        //   error,
+        // );
+
+        await new Promise((resolve) =>
+          setTimeout(resolve, Math.min(delay * 2, this.maxPollingDelay))
+        );
+        delay = Math.min(delay * 2, this.maxPollingDelay);
+        attempts++;
```

Comment:
@sam-goodwin 
do you think the following would be good enough?

I'm not sure how to force any of those errors to properly test this

```suggestion
	if (RETRYABLE_ERRORS.includes(error.message)) {
	        await new Promise((resolve) =>
	          setTimeout(resolve, Math.min(delay * 2, this.maxPollingDelay))
	        );
	        delay = Math.min(delay * 2, this.maxPollingDelay);
	        attempts++;
        }
```

Reply:
Not sure about max attempts? I'd expect the API to eventually fail